samples/matter/common/src/binding_handler.cpp

```diff
@@ -5,83 +5,92 @@
-#include <app/CommandSender.h>
-#include <app/OperationalSessionSetup.h>
-#include <app/clusters/bindings/BindingManager.h>
-#include <app/clusters/bindings/bindings.h>
-#include <app/data-model/Decode.h>
-#include <app/util/binding-table.h>
-#include <controller/InvokeInteraction.h>
-#include <controller/ReadInteraction.h>
-#include <lib/support/CodeUtils.h>
+
+#include <zephyr/logging/log.h>
 
 LOG_MODULE_DECLARE(app, CONFIG_CHIP_APP_LOG_LEVEL);
 
 using namespace chip;
 using namespace chip::app;
 
-void BindingHandler::Init()
+CHIP_ERROR BindingHandler::Init(deviceChangedHandler pHandler)
 {
-	DeviceLayer::PlatformMgr().ScheduleWork(InitInternal);
+	CHIP_ERROR error      = CHIP_NO_ERROR;
+	
+	if(mDeviceChangedHandler != nullptr){
+		error = CHIP_ERROR_INVALID_ARGUMENT;
+	}
+	else{
+		mDeviceChangedHandler = pHandler;
+		DeviceLayer::PlatformMgr().ScheduleWork(InitInternal);
+	}
+
+	return error;
 }
 
-void BindingHandler::TemperatureMeasurementReadHandler(const EmberBindingTableEntry &binding,
-						       OperationalDeviceProxy *peer_device, void *context)
+void BindingHandler::OnInvokeCommandSucces()
 {
-	VerifyOrReturn(context != nullptr, LOG_ERR("Invalid context for Thermostat handler"););
-	BindingData *data = static_cast<BindingData *>(context);
 
-	TemperatureMeasurementUnicastBindingRead(data, binding, peer_device);
-}
+	LOG_DBG("Binding command applied successfully!");
 
-void BindingHandler::TemperatureMeasurementUnicastBindingRead(BindingData *data, const EmberBindingTableEntry &binding,
-							      OperationalDeviceProxy *peer_device)
+	/* If session was recovered and communication works, reset flag to the initial state. */
+	if (BindingHandler::GetInstance().mCaseSessionRecovered)
+		BindingHandler::GetInstance().mCaseSessionRecovered = false;
+}
+void BindingHandler::OnInvokeCommandFailure(BindingData &BindingData, CHIP_ERROR aError)
 {
-	auto onSuccess = [](const ConcreteDataAttributePath &attributePath, const auto &dataResponse) {
-		ChipLogProgress(NotSpecified, "Read Thermostat attribute succeeded");
-		int16_t responseValue = dataResponse.Value();
-		if (!(dataResponse.IsNull())) {
-			TempSensorManager::Instance().SetLocalTemperature(responseValue);
-		}
-	};
+	CHIP_ERROR error;
+
+	if (aError == CHIP_ERROR_TIMEOUT && !BindingHandler::GetInstance().mCaseSessionRecovered) {
+		LOG_INF("Response timeout for invoked command, trying to recover CASE session.");
 
-	auto onFailure = [](const ConcreteDataAttributePath *attributePath, CHIP_ERROR error) {
-		ChipLogError(NotSpecified, "Read Thermostat attribute failed: %" CHIP_ERROR_FORMAT, error.Format());
-	};
+		/* Set flag to not try recover session multiple times. */
+		BindingHandler::GetInstance().mCaseSessionRecovered = true;
 
-	CHIP_ERROR ret = CHIP_NO_ERROR;
+		/* Allocate new object to make sure its life time will be appropriate. */
+		BindingHandler::BindingData *data = Platform::New<BindingHandler::BindingData>();
+		*data = BindingData;
```

Comment:
The `void BindingHandler::OnInvokeCommandFailure(BindingData &BindingData` indicates that the input bindingData must exist when the function is called because it's passed as a reference, is that correct?